File: examples/email-smoke-py/smoke.py
"""End-to-end email-channel smoke test (Python).

Mirror of `examples/email-smoke/smoke.ts` — exercises the same loop
(create task with email notify → capture rendered email → click
magic link → assert agent receives the response) but driven from the
Python SDK so we have parity coverage in both languages.

What runs:

  1. Configure an email sender identity that uses the `file` transport
     (drops one JSON per email into a tmp dir).
  2. Call `await_human` with a single-boolean response schema and
     `notify=["email+<id>:..."]`. The Python SDK's `extract_form`
     produces a Switch primitive, which is what triggers the email
     renderer's magic-link path.
  3. Concurrently poll the tmp dir for the rendered email, parse out
     the Approve magic-link URL, then POST to it.
  4. Wait for `await_human` to resolve. Assert `approved is True`.

Prerequisites (in another terminal):

    awaithumans dev

Then in this terminal:

    cd examples/email-smoke-py
    pip install -r requirements.txt
    export AWAITHUMANS_ADMIN_API_TOKEN="$(cat ~/.awaithumans/admin.token)"
    python smoke.py

The script reads `AWAITHUMANS_URL` (default http://localhost:3001)
and `AWAITHUMANS_ADMIN_API_TOKEN` (required) from the env.
"""

from __future__ import annotations

import asyncio
import json
import os
import re
import sys
import tempfile
import time
from pathlib import Path
from typing import Any

import httpx
from pydantic import BaseModel, Field

from awaithumans import await_human
from awaithumans.utils.discovery import resolve_admin_token, resolve_server_url

# ─── Config ────────────────────────────────────────────────────────────

# Resolve URL + admin token via the same chain `await_human` itself
# uses internally — explicit env var → discovery file written by
# `awaithumans dev`. Means the smoke "just works" against a running
# dev server with no env-var dance, mirroring the Python SDK's
# default DX.
SERVER_URL = resolve_server_url().rstrip("/")
ADMIN_TOKEN = resolve_admin_token()

if not ADMIN_TOKEN:
    print(
        "Couldn't find an admin token. Either:\n"
        "  - Run `awaithumans dev` (writes ~/.awaithumans-dev.json) and try again, OR\n"
        "  - Export AWAITHUMANS_ADMIN_API_TOKEN with the token your server uses.",
        file=sys.stderr,
    )
    sys.exit(1)

RECIPIENT_EMAIL = "smoke-recipient@example.test"
IDENTITY_ID = f"smoke-py-{int(time.time())}"

# Unique per-run tmp dir so old smoke runs don't pollute the magic-link
# search. Slow filesystems would still see "ls then read" race; the
# poller below tolerates it.
EMAIL_DIR = Path(tempfile.mkdtemp(prefix="awaithumans-py-smoke-"))
print(f"→ email capture dir: {EMAIL_DIR}")


# ─── Schemas ───────────────────────────────────────────────────────────


class TransferRequest(BaseModel):
    transfer_id: str = Field(...)
    amount_usd: float
    to: str


class ApprovalResponse(BaseModel):
    """Single boolean → server-side `extract_form` produces a Switch
    primitive, which is what the email renderer uses to decide whether
    to emit Approve/Reject magic-link buttons."""

    approved: bool = Field(..., description="Approve this transfer?")


# ─── Admin helpers ─────────────────────────────────────────────────────


_admin = httpx.Client(
    base_url=SERVER_URL,
    headers={
        "Authorization": f"Bearer {ADMIN_TOKEN}",
        "Content-Type": "application/json",
    },
    timeout=10,
)


def configure_file_transport_identity() -> None:
    resp = _admin.post(
        "/api/channels/email/identities",
        json={
            "id": IDENTITY_ID,
            "display_name": "Smoke test sender",
            "from_email": "smoke@app.example",
            "from_name": "awaithumans py smoke",
            "reply_to": None,
            "transport": "file",
            "transport_config": {"dir": str(EMAIL_DIR)},
        },
    )
    if resp.status_code >= 400:
        raise RuntimeError(
            f"identity create failed: {resp.status_code} {resp.text}"
        )
    print(f"→ created email identity '{IDENTITY_ID}' (file transport)")


def delete_identity() -> None:
    """Best-effort cleanup so a half-run smoke doesn't leave junk in
    the operator's identity list."""
    try:
        _admin.delete(f"/api/channels/email/identities/{IDENTITY_ID}")
    except Exception:  # noqa: BLE001
        # Non-fatal — operator can clear it manually if they care.
        pass


# ─── Magic-link capture ────────────────────────────────────────────────


_ACTION_PATH_RE = re.compile(
    r"/api/channels/email/action/[A-Za-z0-9_\-.]+"
)


async def poll_for_email(deadline: float) -> dict[str, Any]:
    while time.time() < deadline:
        files = sorted(EMAIL_DIR.glob("*.json"))
        for f in files:
            payload = json.loads(f.read_text())
            if payload.get("to") == RECIPIENT_EMAIL:
                return payload
        await asyncio.sleep(0.25)
    raise TimeoutError(
        f"Timed out waiting for email to {RECIPIENT_EMAIL} in {EMAIL_DIR}"
    )


def find_approve_link(email: dict[str, Any]) -> str:
    body_html = email.get("html", "") or ""
    body_text = email.get("text", "") or ""
    match = _ACTION_PATH_RE.search(body_text) or _ACTION_PATH_RE.search(body_html)
    if match is None:
        raise RuntimeError(
            "No magic-link URL found. Was form_definition synthesized? "
            "See packages/python/awaithumans/forms/extract.py.\n"
            f"text body:\n{body_text}\n\nhtml body:\n{body_html}"
        )
    return f"{SERVER_URL}{match.group(0)}"


def click_magic_link(url: str) -> None:
    """Public action endpoint — POST with no body completes the task."""
    resp = httpx.post(url, timeout=10)
    if resp.status_code >= 400:
        raise RuntimeError(
            f"Magic-link POST returned {resp.status_code}: {resp.text}"
        )
    print(f"→ POSTed magic-link → {resp.status_code}")


def assert_email_looks_right(email: dict[str, Any]) -> None:
    if not email.get("subject"):
        raise AssertionError("Email has empty subject")
    html = email.get("html", "")
    if "Approve wire transfer (smoke test)" not in html:
        raise AssertionError("Email body missing the task title")
    if "WT-PYSMOKE-1" not in html:
        raise AssertionError("Email body missing the payload (transfer_id)")


# ─── Orchestration ─────────────────────────────────────────────────────


async def main() -> None:
    print(f"→ smoke against {SERVER_URL}")

    configure_file_transport_identity()

    # `await_human` is awaitable; run it concurrently with the email-
    # capture poll so we can click the magic link while the SDK is
    # still polling.
    agent_task = asyncio.create_task(
        await_human(
            task="Approve wire transfer (smoke test)",
            payload_schema=TransferRequest,
            payload=TransferRequest(
                transfer_id="WT-PYSMOKE-1",
                amount_usd=10_000,
                to="Acme Inc.",
            ),
            response_schema=ApprovalResponse,
            timeout_seconds=300,  # 5 min, plenty of slack
            idempotency_key=f"email-smoke-py-{int(time.time())}",
            notify=[f"email+{IDENTITY_ID}:{RECIPIENT_EMAIL}"],
        )
    )

    # 30s deadline — the notifier runs as a background task after the
    # create-task response, so it always lands within ~1s on a healthy
    # box. 30s is paranoid.
    deadline = time.time() + 30
    email = await poll_for_email(deadline)
    print(f"→ captured email: subject=\"{email['subject']}\" to={email['to']}")
    assert_email_looks_right(email)
    print("→ email body content checks: OK")

    approve_url = find_approve_link(email)
    print(f"→ magic-link URL: {approve_url}")

    click_magic_link(approve_url)

    decision: ApprovalResponse = await agent_task
    if decision.approved is not True:
        raise AssertionError(
            f"Expected approved=True, got: {decision.model_dump()}"
        )

    print("✓ smoke pass: Python SDK + email channel + magic-link round-trip")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    finally:
        delete_identity()
        _admin.close()

File: examples/email-smoke/smoke.ts
/**
 * End-to-end email-channel smoke test (TypeScript).
 *
 * Exercises the full create → notify → click-magic-link → resolve loop
 * against a real awaithumans server, using ONLY the public TypeScript
 * SDK + a tiny admin helper for the email-identity setup. No mocks,
 * no internal imports.
 *
 * What runs:
 *   1. Configure an email sender identity that uses the `file` transport
 *      (drops one JSON per email into a tmp dir — captures what would
 *      otherwise have left for Resend).
 *   2. Call `awaitHuman` with `notify=["email:..."]` and a single-Switch
 *      response schema (the renderer emits magic-link buttons for that).
 *   3. Concurrently poll the tmp dir for the email file, parse the
 *      "Approve" magic-link URL out of the rendered text body, then
 *      POST to it (the action endpoint is public — no auth needed).
 *   4. Wait for `awaitHuman` to resolve. Assert the response matches.
 *
 * Why a runnable script and not a vitest unit test:
 *   - The whole point is "does the TS SDK actually talk to the server."
 *   - Mocked unit coverage already exists (tests/await-human.test.ts).
 *   - This catches contract drift between the wire formats.
 *
 * Prerequisites (in another terminal):
 *
 *     awaithumans dev
 *
 * Then in this terminal:
 *
 *     cd examples/email-smoke
 *     npm install
 *     export AWAITHUMANS_ADMIN_API_TOKEN="$(cat ~/.awaithumans/admin.token)"
 *     npm start
 *
 * The default server URL is http://localhost:3001. Override with
 * AWAITHUMANS_URL if your dev server is elsewhere (e.g., behind ngrok).
 */

import { mkdtempSync, readFileSync, readdirSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { setTimeout as sleep } from "node:timers/promises";

import { awaitHuman } from "awaithumans";
import { z } from "zod";

// ─── Config ────────────────────────────────────────────────────────────

const SERVER_URL =
	process.env.AWAITHUMANS_URL ?? "http://localhost:3001";
const ADMIN_TOKEN = process.env.AWAITHUMANS_ADMIN_API_TOKEN;

if (!ADMIN_TOKEN) {
	console.error(
		"AWAITHUMANS_ADMIN_API_TOKEN is required. " +
			"Run `awaithumans dev` first, then set the env var to the contents " +
			"of ~/.awaithumans/admin.token (or wherever your dev DB sits).",
	);
	process.exit(1);
}

const RECIPIENT_EMAIL = "smoke-recipient@example.test";
const IDENTITY_ID = `smoke-${Date.now().toString(36)}`;

// Use a unique tmp dir per run so old emails from prior smoke runs
// don't pollute the magic-link search.
const EMAIL_DIR = mkdtempSync(join(tmpdir(), "awaithumans-smoke-"));
console.log(`→ email capture dir: ${EMAIL_DIR}`);

// ─── Schemas ───────────────────────────────────────────────────────────

const TransferRequest = z.object({
	transferId: z.string(),
	amountUsd: z.number(),
	to: z.string(),
});

// IMPORTANT: a single boolean field triggers the renderer's magic-
// link path (Approve/Reject buttons in the email). Multi-field
// responses fall back to a "review in dashboard" link-out, which
// is harder to script.
const ApprovalResponse = z.object({
	approved: z.boolean().describe("Approve this transfer?"),
});

// ─── Admin helpers ─────────────────────────────────────────────────────

async function adminFetch(
	path: string,
	init: RequestInit = {},
): Promise<Response> {
	return fetch(`${SERVER_URL}${path}`, {
		...init,
		headers: {
			...(init.headers ?? {}),
			Authorization: `Bearer ${ADMIN_TOKEN}`,
			"Content-Type": "application/json",
		},
	});
}

async function configureFileTransportIdentity(): Promise<void> {
	const resp = await adminFetch("/api/channels/email/identities", {
		method: "POST",
		body: JSON.stringify({
			id: IDENTITY_ID,
			display_name: "Smoke test sender",
			from_email: "smoke@app.example",
			from_name: "awaithumans smoke",
			reply_to: null,
			transport: "file",
			transport_config: { dir: EMAIL_DIR },
		}),
	});
	if (!resp.ok) {
		throw new Error(
			`identity create failed: ${resp.status} ${await resp.text()}`,
		);
	}
	console.log(`→ created email identity '${IDENTITY_ID}' (file transport)`);
}

async function deleteIdentity(): Promise<void> {
	// Best-effort cleanup so a half-run smoke test doesn't leave junk
	// rows in the operator's email-identity list.
	try {
		await adminFetch(`/api/channels/email/identities/${IDENTITY_ID}`, {
			method: "DELETE",
		});
	} catch {
		// non-fatal — the row is harmless and the operator can clear
		// it manually if they care.
	}
}

// ─── Email capture ────────────────────────────────────────────────────

interface CapturedEmail {
	to: string;
	subject: string;
	html: string;
	text: string;
}

async function pollForEmail(
	deadlineMs: number,
): Promise<CapturedEmail> {
	while (Date.now() < deadlineMs) {
		const files = readdirSync(EMAIL_DIR)
			.filter((f) => f.endsWith(".json"))
			.sort(); // unix-ms-prefixed names sort chronologically

		const recent = files.find((f: string) => {
			const payload = JSON.parse(
				readFileSync(join(EMAIL_DIR, f), "utf8"),
			);
			return payload.to === RECIPIENT_EMAIL;
		});

		if (recent) {
			const payload = JSON.parse(
				readFileSync(join(EMAIL_DIR, recent), "utf8"),
			);
			return {
				to: payload.to,
				subject: payload.subject,
				html: payload.html,
				text: payload.text,
			};
		}
		await sleep(250);
	}
	throw new Error(
		`Timed out waiting for email to ${RECIPIENT_EMAIL} in ${EMAIL_DIR}`,
	);
}

function assertEmailLooksRight(email: CapturedEmail): void {
	// Pin the things the email channel SHOULD always include — anything
	// missing here would mean the renderer regressed.
	if (!email.subject) {
		throw new Error("Email has empty subject");
	}
	if (!email.html.includes("Approve wire transfer (smoke test)")) {
		throw new Error("Email body missing the task title");
	}
	if (!email.html.includes("WT-SMOKE-1")) {
		throw new Error("Email body missing the payload (transferId)");
	}
	if (!email.html.includes(`/tasks/`) && !email.html.includes(`/task?id=`)) {
		throw new Error("Email body missing the dashboard link-out");
	}
}

async function completeTaskViaAdmin(taskId: string): Promise<void> {
	// `awaitHuman` doesn't currently synthesize a `form_definition` from
	// a Zod schema (Python has `extract_form`; the TS port is a
	// post-launch task). Without that, the email renderer falls back
	// to a "Review in dashboard" link-out — no magic-link buttons. So
	// the smoke test completes the task via the admin completion API
	// instead of clicking a magic link.
	//
	// What this still proves end-to-end:
	//   - TS SDK's `awaitHuman` creates the task with the right wire shape
	//   - The notify list resolves to the email channel
	//   - The email transport actually fires + writes the message
	//   - Polling resolves when the task transitions to completed
	//
	// The magic-link click path has its own Python coverage; once form
	// synthesis lands in the TS SDK we'll switch this script to click
	// the link instead.
	const resp = await adminFetch(`/api/tasks/${taskId}/complete`, {
		method: "POST",
		body: JSON.stringify({
			response: { approved: true },
			completed_via_channel: "smoke",
		}),
	});
	if (!resp.ok) {
		throw new Error(
			`task complete failed: ${resp.status} ${await resp.text()}`,
		);
	}
	console.log(`→ completed task ${taskId} via admin API`);
}

// ─── Orchestration ─────────────────────────────────────────────────────

async function main(): Promise<void> {
	console.log(`→ smoke against ${SERVER_URL}`);

	await configureFileTransportIdentity();

	// `awaitHuman` is an awaitable that resolves on completion. Run it
	// concurrently with the email-capture poll so we can click the
	// magic link while the SDK is still polling.
	const idem = `email-smoke-${Date.now()}`;

	const awaitPromise = awaitHuman({
		task: "Approve wire transfer (smoke test)",
		payloadSchema: TransferRequest,
		payload: {
			transferId: "WT-SMOKE-1",
			amountUsd: 10_000,
			to: "Acme Inc.",
		},
		responseSchema: ApprovalResponse,
		timeoutMs: 5 * 60_000, // 5 minutes — plenty of slack for CI
		idempotencyKey: idem,
		// notify format: `<channel>[+<identity>]:<target>`. The `+identity`
		// suffix on the channel side picks our smoke-test sender; without
		// it the notifier would fall back to env-configured defaults.
		notify: [`email+${IDENTITY_ID}:${RECIPIENT_EMAIL}`],
		serverUrl: SERVER_URL,
		apiKey: ADMIN_TOKEN,
	});

	// 30s deadline for the background email — the notifier runs as a
	// FastAPI BackgroundTask AFTER the create-task response, so it
	// always lands within ~1s on a healthy box. 30s is paranoid.
	// 30s deadline for the background email — the notifier runs as a
	// FastAPI BackgroundTask AFTER the create-task response, so it
	// always lands within ~1s on a healthy box. 30s is paranoid.
	const captureDeadline = Date.now() + 30_000;
	const email = await pollForEmail(captureDeadline);
	console.log(`→ captured email: subject="${email.subject}" to=${email.to}`);
	assertEmailLooksRight(email);
	console.log("→ email body content checks: OK");

	// Find the task we just created so we can complete it via admin API.
	// The list endpoint returns most-recent-first; we filter by our
	// idempotency_key for an exact match.
	const taskId = await findTaskByIdempotencyKey(idem);
	console.log(`→ resolved task_id=${taskId}`);

	await completeTaskViaAdmin(taskId);

	const decision = await awaitPromise;
	if (decision.approved !== true) {
		throw new Error(
			`Expected approved=true, got: ${JSON.stringify(decision)}`,
		);
	}

	console.log("✓ smoke pass: TS SDK created task → email captured → SDK polled → resolved");
}

async function findTaskByIdempotencyKey(idem: string): Promise<string> {
	// The TS SDK doesn't yet expose the task_id from awaitHuman directly
	// (it returns the typed response, not the wire record). So the
	// smoke test recovers the id via the admin list — fine for a
	// dev-mode test, not a pattern we'd recommend for production code.
	const resp = await adminFetch("/api/tasks?limit=50");
	if (!resp.ok) {
		throw new Error(`task list failed: ${resp.status}`);
	}
	const tasks = (await resp.json()) as Array<{
		id: string;
		idempotency_key: string;
	}>;
	const match = tasks.find((t) => t.idempotency_key === idem);
	if (!match) {
		throw new Error(
			`Couldn't find task with idempotency_key=${idem} in the listing`,
		);
	}
	return match.id;
}

main()
	.catch(async (err) => {
		console.error("✗ smoke FAILED");
		console.error(err);
		await deleteIdentity();
		process.exit(1);
	})
	.then(async () => {
		await deleteIdentity();
	});
